### PR TITLE
Make EventFd non-blocking

### LINKED ISSUE #129: Add Try_Read Functionality to EventFd Implementation
We are currently using EventFds via the implementation from the sys_util crate, inherited from crosvm. One potentially significant downside is that it only exposes a blocking `read()` API. 

There are places in our code where we use `read()` knowing that an event should definitely be available, because the logic is programmed that way, but if bugs are introduces, or something goes terribly wrong otherwise, the code will simply hang on the `read()` (not even panic).

Maybe it would make sense to add some sort of `try_read()` functionality, and do checked reads in this context. There will be some additional overhead involved, but it should not be significant.

### PR DESCRIPTION
**Pull request by lloydmeta**
Closes #129

Description of changes:

* Make EventFd non-blocking by passing `EFD_NONBLOCK` on eventfd creation

### CODE REVIEW

File: sys_util/src/eventfd.rs [hunk 1]
```diff
@@ -5,10 +5,11 @@
 use std::fs::File;
 use std::mem;
 use std::os::unix::io::{AsRawFd, FromRawFd, RawFd};
+use std::result;
 
-use libc::{c_void, dup, eventfd, read, write};
+use libc::{c_int, c_void, dup, eventfd, poll, pollfd, read, write, POLLIN};
 
-use {errno_result, Result};
+use {errno_result, Error as ErrNoErr, Result};
```
[the remark this answers is not in the document]
Comment (lloydmeta) [reply sys_util/src/eventfd.rs]: Good idea. I've gone and made changes in fb28efa to stop using `sys_util::Error` in `EventFd` methods in the spirit of #858.

File: sys_util/src/eventfd.rs [hunk 2]
```diff
@@ -91,6 +109,12 @@ impl AsRawFd for EventFd {
     }
 }
 
+#[derive(Debug)]
+pub enum Error {
```
[the remark this answers is not in the document]
Comment (lloydmeta) [reply sys_util/src/eventfd.rs]: Done.

File: sys_util/src/eventfd.rs [hunk 3]
```diff
@@ -69,6 +70,23 @@ impl EventFd {
         Ok(buf)
     }
 
+    pub fn try_read(&self) -> result::Result<u64, Error> {
```
[the remark this answers is not in the document]
Comment (lloydmeta) [reply sys_util/src/eventfd.rs]: Done.

File: vmm/src/device_manager/mmio.rs [hunk 4]
```diff
@@ -312,10 +313,22 @@ mod tests {
...
         );
-        let e = Error::CloneIoeventFd(sys_util::Error::new(0));
-        assert_eq!(format!("{}", e), "failed to clone ioeventfd: Error(0)");
-        let e = Error::CloneIrqFd(sys_util::Error::new(0));
-        assert_eq!(format!("{}", e), "failed to clone irqfd: Error(0)");
+        let e = Error::CloneIoeventFd(io::Error::from_raw_os_error(0));
+        assert_eq!(
+            format!("{}", e),
+            format!(
+                "failed to clone ioeventfd: {:?}",
```
Comment (aghecenco) [inline vmm/src/device_manager/mmio.rs]: Use the actual error text and get rid of the `format!`:

```suggestion
                "failed to clone ioeventfd: No error information (os error 0)",
```

File: vmm/src/device_manager/mmio.rs [hunk 5]
```diff
@@ -312,10 +313,22 @@ mod tests {
...
+                io::Error::from_raw_os_error(0)
+            )
+        );
+        let e = Error::CloneIrqFd(io::Error::from_raw_os_error(0));
+        assert_eq!(
+            format!("{}", e),
+            format!(
+                "failed to clone irqfd: {:?}",
```
Comment (aghecenco) [inline vmm/src/device_manager/mmio.rs]:
```suggestion
                "failed to clone irqfd: No error information (os error 0)",
```
and remove the `format!`

File: vmm/src/device_manager/mmio.rs (hunk 4 above)
Comment (lloydmeta) [reply vmm/src/device_manager/mmio.rs]: Looks like this won't work exactly the same way between gnu and musl (maybe even others).

I'll revert this to using format by `{}` instead for now.

```rust
test result: ok. 23 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running /tmp/firecracker_test_session/firecracker_binaries/test/x86_64-unknown-linux-gnu/debug/deps/vmm-b33a6f9ca44b473a

running 45 tests
// snip
test device_manager::mmio::tests::test_error_messages ... FAILED
// snip

     Running /tmp/firecracker_test_session/firecracker_binaries/test/x86_64-unknown-linux-musl/debug/deps/vmm-b33a6f9ca44b473a

running 47 tests
// snip
test device_manager::mmio::tests::test_error_messages ... ok
// snip
```

**Comment by acatangiu** [comment]
Will review soon, as a first comment I'm not sure `poll()` is the way to go to implement this.

**Comment by lloydmeta** [comment]
> @lloydmeta thanks for your contribution! The changes in this PR are all in the right direction, but some changes are needed on the implementation of the `EventFd#try_read`.
> 
> Using `poll()` is inefficient in this context. I suggest using the builtin `eventfd` blocking/non-blocking capabilities.
> 
> The way I see it there are two ways to go:
> 
> 1. Change the existing `EventFd` object implementation to create the underlying `eventfd` using the `EFD_NONBLOCK` flag, thus making `EventFd#read` non-blocking. This option is easier but less flexible (cannot use blocking `read()`s anymore).
> 2. Create a new class `EventFdNonBlock` - or some similar name - and implement all common functionality between the two classes as traits, the only difference between them being the constructor which in one case uses `EFD_NONBLOCK` and the other does not.
> 
> Option 2 is more flexible, but I don't know if we really need that flexibility. Do we ever want to use blocking `read()`s on `EventFd`?

Good points, there is definitely some overhead introduced using `poll` due to the check-and-read mechanism.  Using non-blocking on the FD would definitely get rid of it, but it seems to be a different ask than that which was described in #129, which I read to be to *add* a `try_read()`, and that some overhead would be ok ("There will be some additional overhead involved, but it should not be significant"). To be clear, I'm fine with re-working the implementation, but I just wanted to make sure we're on the same page wrt reasoning.

To enable the use of non-blocking mode on an FD, [as you said](https://github.com/firecracker-microvm/firecracker/pull/863#discussion_r250348742), there are two approaches:

1. `EFD_NONBLOCK` on `eventfd` creation:
   * This will also make writes non-blocking ... which is not-backwards compatible because the behaviour changes so that we have another error case to handle , making the change set bigger, depending on how you want to handle that case (ignore or further wrap and panic again?)
2. `fcntl` before `try_read()` and unsetting it afterwards (back to blocking)
    *  This is backwards compatible on write
    * ...but setting and unsetting the mode adds additional overhead anyways.

Which one do we want to go with? For me, the the decision comes down to whether it makes sense to make `write` non-blocking as well.  It seems to me for `read()`, it makes sense that we might want to completely switch to non-blocking (though, again, a different ask vs #129 which was to add a non-blocking method), since it is more error prone; `write()` on the other hand seems less of a candidate, given the actual error case is, based on [the man page](http://man7.org/linux/man-pages/man2/eventfd.2.html), ... pretty unlikely:

>               The maximum value that may be stored
>               in the counter is the largest unsigned 64-bit value minus 1
>               (i.e., 0xfffffffffffffffe).  If the addition would cause the
>               counter's value to exceed the maximum, then the write(2)
>               either blocks until a read(2) is performed on the file
>               descriptor, or fails with the error EAGAIN if the file
>               descriptor has been made nonblocking.

> I like the `sys_util::error` replacements, those commits look good and are not tied to the `EventFd` implementation. You can move them to a separate PR so we can merge them sooner.

Agreed; that part was a bit out of scope for #129 anyways. Pulled it out and put up another PR #892

**Comment by acatangiu** [comment]
@lloydmeta #129 should not be taken word-for-word, it is meant to track that we want a non-blocking mechanism for event-fds.

I discussed with @alexandruag - who originally opened #129 - and agreed that the best and easiest way to go about this is to change the event-fd constructor to use `EFD_NONBLOCK` on the `eventfd()` call/creation.

This is indeed a breaking change but the new `Error` for the `EAGAIN` case should be handled in the same way as the other errors.

**Comment by lloydmeta** [comment]
> @lloydmeta #129 should not be taken word-for-word, it is meant to track that we want a non-blocking mechanism for event-fds.
> 
> I discussed with @alexandruag - who originally opened #129 - and agreed that the best and easiest way to go about this is to change the event-fd constructor to use `EFD_NONBLOCK` on the `eventfd()` call/creation.
> 
> This is indeed a breaking change but the new `Error` for the `EAGAIN` case should be handled in the same way as the other errors.

Yep, sounds reasonable to me; just wanted to make sure we were all on the same page wrt what the goal was, reasoning, and pros-and-cons. It sounds like we're settled on making the eventfd non-blocking, which is fine by me :)

To recap:

1. Use non blocking; and thus make both `read()` and `write()` non-blocking
2. `EAGAIN` would be returned from the system via `io::Error::last_os_error()` and thus treated the same as the other errors (check for return < 0 for both [read()](http://man7.org/linux/man-pages/man2/read.2.html) and [write()](http://man7.org/linux/man-pages/man2/write.2.html) syscalls)
3. We can remove the added sys call whitelist for poll

Does that sound about right ?

It also feels like this would be nicer to work on once https://github.com/firecracker-microvm/firecracker/pull/892 is merged, just to get rid of the noise.

**Comment by acatangiu** [comment]
@lloydmeta agreed on all counts :+1:

**Pull request merged**